operatorcourier/flatten.py

```diff
@@ -0,0 +1,151 @@
+    :param version_folder_name: The name of the version folder containing bundle files
+    :param csv_paths: A list of CSV file paths inside version folders
+    :param crd_dict: dict that contains CRD info collected from different version folders,
+    where the key is the CRD name, and the value is a tuple where the first element is
+    the version of the bundle, and the second is the path of the CRD file
+    """
+    # parse each version folder and parse CRD, CSV files
+    try:
+        semver.parse(version_folder_name)
+    except ValueError:
+        logger.warning("Ignoring %s as it is not a valid semver. "
+                       "See https://semver.org for the semver specification.",
+                       version_folder_name)
+        return
+
+    logger.info('Parsing folder: %s...', version_folder_name)
+
+    contains_csv = False
+    for item in os.listdir(os.path.join(base_dir, version_folder_name)):
+        version_folder_path = os.path.join(base_dir, version_folder_name)
+        item_path = os.path.join(version_folder_path, item)
+
+        if not os.path.isfile(item_path):
+            logger.warning('Ignoring %s as it is not a regular file.', item)
+
+        with open(item_path, 'r') as f:
+            file_content = f.read()
+
+        yaml_type = identify.get_operator_artifact_type(file_content)
+
+        if yaml_type == 'ClusterServiceVersion':
+            contains_csv = True
+            csv_paths.append(item_path)
+        elif yaml_type == 'CustomResourceDefinition':
+            try:
+                crd_name = safe_load(file_content)['metadata']['name']
+            except MarkedYAMLError:
+                msg = "Courier requires valid input YAML files"
+                logger.error(msg)
+                raise errors.OpCourierBadYaml(msg)
+            # create new CRD type entry if not found in dict
+            if crd_name not in crd_dict:
+                crd_dict[crd_name] = (version_folder_name, item_path)
+            # update the CRD type entry with the file with the newest version
+            elif semver.compare(version_folder_name, crd_dict[crd_name][0]) == 1:
+                crd_dict[crd_name] = (crd_dict[crd_name][0], item_path)
+
+    if not contains_csv:
+        msg = 'This version directory does not contain any valid CSV file.'
+        logger.error(msg)
+        raise errors.OpCourierBadBundle(msg)
+
+
+def get_package_file(base_dir: str, file_names_in_base_dir) -> (str, str):
+    packages = []  # [ (PACKAGE_FILE_PATH, PACKAGE_FILE_NAME) ]
+
+    # add package file to file_paths_to_copy
+    # only 1 package yaml file is expected in file_names
+    for file_name in file_names_in_base_dir:
+        file_path = os.path.join(base_dir, file_name)
+        with open(file_path, 'r') as f:
+            file_content = f.read()
+        if identify.get_operator_artifact_type(file_content) != 'Package':
+            logger.warning('Ignoring %s as it is not a valid package file.', file_name)
```

Comment:
Directory should be used too, is hard to identify which file when files are nested in dirs.

Reply:
@MartinBasti Do you mean that I need to recursively search for package files in the source_dir? I think we should only expect the package file to exist in the root directory and ignoring those in version folders.

Reply:
If we assume the package is in the root directory then don't we not actually need to iterate over them?

Reply:
@kevinrizza Yes iterating over the root directory is done [here](https://github.com/operator-framework/operator-courier/blob/2a9a333bcd/operatorcourier/flatten.py#L23) (but only for 1 level) , and the input of this helper is expected be only files coming from the root dir, as is indicated by its arg name `file_names_in_base_dir`.

Reply:
ok, if this is root dir I'm fine, I was afraid that it uses subdirectories.